### PR TITLE
Introduce InteractionParticle

### PR DESCRIPTION
**Pull request by TUsama**
Including particles when the player deals damage, entity dodge and resist, and upcoming healing number particle

### CODE REVIEW

**Comment by RobertSkalko** [comment]
I'm a bit confused

1) Why is the event created? Why not just straight send packet?

2) Why the many types of particles, why not just 1? A dmg particle is just colored text

**Comment by TUsama** [comment]
1. To separate the code logic further. This can keep the DamageEvent class clean. The whole system is not just for particle generation but also for the whole project. it is very useful when some logic needs to be repeatedly called and you can see this logic would be changed in the future. In this case, if we need to send the particle to everyone(assuming), we can just change the code in the listener class, otherwise, we need to change 3.
2. I will add a Wynncraft-like style because it's clean and easy to read at some point. That's also why I need to separate the spawn logic and render logic.

**Comment by RobertSkalko** [comment]
What do you mean keep the class clean? What's the difference between sending a packet or calling some method to do it, and calling an event? Just get rid of the event. Event here would only be useful if you want to allow addon mods to do something on the event somehow. Either way i'd rather not use forge's events for this

**Comment by TUsama** [comment]
First of all, this is not Forge's event system, it's directly from Guava. Forge's event also inherits from it. It won't hinder the version migration at all unless Minecraft gives up Guava in the future.
Second, Essentially there is no difference in code function implementation without using the event. I just want to avoid the situation I mentioned above: what if we need to change this part of logic? What if we need to further expand the particle type? It would be harder and harder to maintain the correspondence between the particle generation component and other function components if we introduce more damage features. My purpose is not to make this project more chunky or show off my code technique, I just want to keep the code easy to be changed. Currently we have 3 places that will call the particle generation, if we can guarantee we won't add any of the particle types and change the packet send logic, I'm glad to delete the event system because I have to say the event thing is a little chunky, comparing the simple method implementation.

**Comment by RobertSkalko** [comment]
Honestly still don't understand but this at least doesn't edit much of the code so it's not risky to merge.  But why did you just add some selena thing from curse maven?

**Comment by TUsama** [comment]
for the test, I'll clean it up when this PR is ready

**Comment by TUsama** [comment]
I think this PR is ready to be reviewd

**Comment by RobertSkalko** [comment]
I still see cursemaven as extra maven

**Comment by RobertSkalko** [comment]
Why does this modify healthbar renderer with boss tags

**Comment by TUsama** [comment]
done

**Comment by RobertSkalko** [comment]
Alright gonna merge to check it out

File: src/main/java/com/robertx22/mine_and_slash/a_libraries/dmg_number_particle/particle/InteractionResultHandler.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+import java.util.Random;
+import java.util.function.BiConsumer;
+
+public class InteractionResultHandler {
+
+
+
+    public enum ParticleSpawnType {
+        DAMAGE(ClientConfigs.getConfig().DAMAGE_PARTICLE_STYLE.get().damageStrategy, new ElementDamageParticle.DamageInformation(null, null, false)),
+        NULLIFIED_DAMAGE(ClientConfigs.getConfig().DAMAGE_PARTICLE_STYLE.get().nullifiedDamageStrategy, DamageNullifiedParticle.Type.DODGE),
```
Comment (RobertSkalko) [inline src/main/java/com/robertx22/mine_and_slash/a_libraries/dmg_number_particle/particle/InteractionResultHandler.java:30]: this part seems bad. Do not save config values and especially dont call them in constructors

**Comment by RobertSkalko** [comment]
Seems good besides that, maybe just rename the DAMAGE_PARTICLE_STYLE config to something better.. original and row i honestly dont know what means at all.

**Comment by RobertSkalko** [comment]
Did you check if it works on servers btw?

**Comment by TUsama** [comment]
> Did you check if it works on servers btw?

I don't have a server to test it currently

**Comment by RobertSkalko** [comment]
You use IDE and click on runserver

**Comment by TUsama** [comment]
> Seems good besides that, maybe just rename the DAMAGE_PARTICLE_STYLE config to something better.. original and row i honestly dont know what means at all.

need me to open a new PR?

**Comment by RobertSkalko** [comment]
Yup open new PR. and yes it crashes on server, seems you use client classes on the server side..
[latest.log](https://github.com/user-attachments/files/17554622/latest.log)

**Comment by RobertSkalko** [comment]
So it crashes servers and disconnects the client. That's why you need to test stuff on servers

**Comment by RobertSkalko** [comment]
From a quick look, you called the client config from damage event. Don't do that. Send the player particle data and let the client decide how to display it

**Comment by TUsama** [comment]
I'll handle it on the next PR, for now just reset that for safety.
First time know IDE can run server directly...

**Comment by RobertSkalko** [comment]
What IDE are you using? And sure i'll revert the PR for now

**Comment by TUsama** [comment]
Jetbrain, I just skip it every time when I see the runServer and click the runClient...

**Comment by RobertSkalko** [comment]
Ah it's in that button, run both client and server, make sure the server has eula true, and allow offline etc

**Comment by TUsama** [comment]
> Seems good besides that, maybe just rename the DAMAGE_PARTICLE_STYLE config to something better.. original and row i honestly dont know what means at all.

how about SINGLE and IN_A_ROW?

**Comment by RobertSkalko** [comment]
Maybe DEFAULT and IN_A_ROW.

**Comment by RobertSkalko** [comment]
Also, i didn't test how the particles show when you have lots of added damage types. Like a fireball with added frost, added chaos dmg etc support gems

**Pull request merged**